### PR TITLE
Downgrade support

### PR DESCRIPTION
**Pull request by abhi16394**
# Description
CSM-Operator is able to roll back updates when the user updates the versions in the csm object to an older version.
Also fixed a bug in CSM-operator when the user downgrades to a version , the config version annotation of csm-object would remain the same and wouldn't change with the downgrade. 
Fixed it by calling the method which would set the annotation depending on the ConfigVersion of CR applied during downgrade
unit-test and e2e test will be added as a separate PR

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- Installed powerflex -2.10.0 version 

image

- Downgraded to powerflex - 2.8.0 version by editing the csm-object image and configversion 
[root@master-1-Zaglt7mQUY8Wg csm-operator]# kubectl get csm -A
NAMESPACE   NAME       CREATIONTIME   CSIDRIVERTYPE   CONFIGVERSION   STATE
vxflexos    vxflexos   3d21h          powerflex       v2.8.0          Succeeded

image

### CODE REVIEW

**Review by bjiang27 — DISMISSED** [review]

**Review by AkshaySainiDell — CHANGES_REQUESTED** [review]

**Comment by AkshaySainiDell** [comment]
@abhi16394 I didn't see changes for validating downgrade. Is this already supported ?

**Comment by abhi16394** [comment]
> @abhi16394 I didn't see changes for validating downgrade. Is this already supported ?

it was supported by upgrade code but made some changes and have added few code after i found a bug in version check

File: controllers/csm_controller.go [hunk 1]
```diff
@@ -1407,6 +1409,16 @@ func applyConfigVersionAnnotations(ctx context.Context, instance *csmv1.Containe
...
+	} else {
+		if instance.HasModule(csmv1.AuthorizationServer) {
+			configVersion = instance.GetModule(csmv1.AuthorizationServer).ConfigVersion
+		} else {
+			configVersion = instance.Spec.Driver.ConfigVersion
+		}
+		annotations[configVersionKey] = configVersion
+		isUpdated = true
+		log.Infof("Installing csm component %s with config Version %s. Updating Annotations with Config Version",
+			instance.GetName(), configVersion)
```
[the remark this answers is not in the document]
Comment (abhi16394) [reply controllers/csm_controller.go]: changed it

File: pkg/utils/utils.go [hunk 2]
```diff
@@ -1126,22 +1128,37 @@ func IsValidUpgrade[T csmv1.CSMComponentType](ctx context.Context, oldVersion, n
...
-	// if not equal, it is an upgrade/downgrade
-	// get minimum required version for upgrade
-	minUpgradePath, err := getUpgradeInfo(ctx, operatorConfig, csmComponentType, newVersion)
+	var minUpgradePath string
+	var err error
+	var isUpgradeValid bool
+
+	log.Info("####oldVersion: ", oldVersion, " ###newVersion: ", newVersion)
+
+	isVersionCheck, _ := MinVersionCheck(oldVersion, newVersion)
```
Comment (AkshaySainiDell) [inline pkg/utils/utils.go:1137]:
```suggestion
	isUpgrade, _ := MinVersionCheck(oldVersion, newVersion)
```
Apart from this, what do you think about renaming downgrade related variables e.g `isDowngradeValid`,  `maxDowngradePath`/`minDowngradePath`

**Review by AkshaySainiDell — CHANGES_REQUESTED** [review]

File: pkg/utils/utils.go (hunk 2 above)
Comment (abhi16394) [reply pkg/utils/utils.go:1137]: i guess isversioncheck should be good as its just veryfying if it should be upgrade or downgrade

Comment (abhi16394) [reply pkg/utils/utils.go]: changed the variable names

**Review by AkshaySainiDell — APPROVED** [review]
lgtm

**Review by bjiang27 — APPROVED** [review]

**Review by kumarkgosa — APPROVED** [review]

**Comment by abhi16394** [comment]
Raised another PR to fix the unsigned commits issue : https://github.com/dell/csm-operator/pull/584

**Pull request closed without merging**